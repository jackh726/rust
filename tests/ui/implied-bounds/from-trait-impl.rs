// check-pass

trait Trait {
    type Assoc;
}

impl<X: 'static> Trait for (X,) {
    type Assoc = ();
}

struct Foo<T: Trait>(T)
where
    T::Assoc: Clone; // any predicate using `T::Assoc` works here

fn func1(foo: Foo<(&str,)>) {
    //~^ WARN function is missing necessary lifetime bounds
    //~| WARN this was previously accepted
    let _: &'static str = foo.0.0;
}

trait TestTrait {}

impl<X> TestTrait for [Foo<(X,)>; 1] {}
//~^ WARN implementation is missing necessary lifetime bounds
//~| WARN this was previously accepted

fn main() {}
