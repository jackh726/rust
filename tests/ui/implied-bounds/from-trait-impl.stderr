warning: function is missing necessary lifetime bounds
  --> $DIR/from-trait-impl.rs:15:1
   |
LL | fn func1(foo: Foo<(&str,)>) {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #109628 <https://github.com/rust-lang/rust/issues/109628>
   = note: to get more detailed errors, use `#[deny(implied_bounds_from_trait_impl)]`
   = note: `#[warn(implied_bounds_from_trait_impl)]` on by default

warning: implementation is missing necessary lifetime bounds
  --> $DIR/from-trait-impl.rs:23:1
   |
LL | impl<X> TestTrait for [Foo<(X,)>; 1] {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #109628 <https://github.com/rust-lang/rust/issues/109628>
   = note: to get more detailed errors, use `#[deny(implied_bounds_from_trait_impl)]`

warning: 2 warnings emitted

